packages/editor/src/store/actions.js

```diff
@@ -26,22 +26,58 @@ import {
 } from './utils/notice-builder';
 
 /**
- * Returns an action object used in signalling that editor has initialized with
+ * WordPress dependencies
+ */
+import {
+	parse,
+	synchronizeBlocksWithTemplate,
+} from '@wordpress/blocks';
+
+/**
+ * Returns an action generator used in signalling that editor has initialized with
  * the specified post object and editor settings.
  *
  * @param {Object} post      Post object.
  * @param {Object} edits     Initial edited attributes object.
  * @param {Array?} template  Block Template.
- *
- * @return {Object} Action object.
  */
-export function setupEditor( post, edits, template ) {
-	return {
+export function* setupEditor( post, edits, template ) {
+	yield {
 		type: 'SETUP_EDITOR',
 		post,
 		edits,
 		template,
 	};
+
+	// In order to ensure maximum of a single parse during setup, edits are
+	// included as part of editor setup action. Assume edited content as
+	// canonical if provided, falling back to post.
+	let content;
+	if ( has( edits, [ 'content' ] ) ) {
+		content = edits.content;
+	} else {
+		content = post.content.raw;
+	}
+
+	let blocks = parse( content );
+
+	// Apply a template for new posts only, if exists.
+	const isNewPost = post.status === 'auto-draft';
+	if ( isNewPost && template ) {
+		blocks = synchronizeBlocksWithTemplate( blocks, template );
+	}
+
+	yield dispatch(
```

Comment:
I can't recall if this had come up in the discussion #13716, but is there any reason we couldn't just yield the result of the action directly?

i.e.

```js
yield resetEditorBlocks( blocks );
```

Same note for `setupEditorState` below.

Reply:
It's mostly just evolved from "consistently" yielding control actions.  But since then I've realized that yielding action object is probably fine (as I did in a few cases in this pull).  So ya, there's probably no need here to yield the controls when the dispatched action is not a generator itself.  I'll update this pull.